lib.rs

```diff
@@ -0,0 +1,280 @@
+        asset_token: contract_ref!(PSP22),
+        payment_token: contract_ref!(PSP22),
+        start_price: u128,
+        min_price: u128,
+        start_time: BlockNumber,
+        end_time: BlockNumber,
+    }
+
+    #[derive(Eq, PartialEq, Debug, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum Error {
+        PSP22TokenCall(PSP22Error),
+        MaxPriceExceeded,
+        InsufficientSupplyToken,
+        NotAuctionOwner,
+    }
+
+    #[ink(event)]
+    #[derive(Clone, Eq, PartialEq, Debug)]
+    pub struct AssetBought {
+        #[ink(topic)]
+        pub by: AccountId,
+        pub price: u128,
+        pub amount: u128,
+    }
+
+    impl From<PSP22Error> for Error {
+        fn from(inner: PSP22Error) -> Self {
+            Error::PSP22TokenCall(inner)
+        }
+    }
+
+    impl DutchAuction {
+        /// Constructor that initializes the contract storrage.
+        ///
+        /// Caller would be the auction_owner
+        #[ink(constructor)]
+        pub fn new(
+            asset_token: AccountId,
+            payment_token: AccountId,
+            start_price: u128,
+            min_price: u128,
+            end_time: BlockNumber,
+        ) -> Self {
+            Self {
+                auction_owner: Self::env().caller(),
+                asset_token: asset_token.into(),
+                payment_token: payment_token.into(),
+                start_price,
+                min_price,
+                start_time: Self::env().block_number(),
+                end_time,
+            }
+        }
+
+        /// The block after which the price will no longer decrease.
+        ///
+        /// The contract will decrease the price linearly from start_price()
+        /// to `min_price()` over the period from `start_time()` to 'end_time()`.
+        /// The auction doesn't end after the period elapses -
+        /// the asset remains available for purchase at `min_price()`.
+        #[ink(message)]
+        pub fn end_time(&self) -> BlockNumber {
+            self.end_time
+        }
+
+        /// The block at which the auction starts
+        #[ink(message)]
+        pub fn start_block(&self) -> BlockNumber {
+            self.start_time
+        }
+
+        /// The price the contract would charge when buying at the current block.
+        #[ink(message)]
+        pub fn price(&self) -> u128 {
+            self.current_price()
+        }
+
+        /// Amount of tokens available for sale.
+        #[ink(message)]
+        pub fn available_asset(&self) -> u128 {
+            self.asset_balance()
+        }
+
+        /// The minimal price the contract allows.
+        #[ink(message)]
+        pub fn min_price(&self) -> u128 {
+            self.min_price
+        }
+
+        /// Update the minimal price.
+        ///
+        /// Requires auction_owner to execute.
+        #[ink(message)]
+        pub fn set_min_price(&mut self, value: u128) -> Result<(), Error> {
```

Comment:
Not sure about having `set_min_price` and `set_end_time` once an auction has started; it depends on specs but we probably want to keep it permissionless.

Reply:
These are essential parameters for calculating the price and controlling the duration of auction. I believe it is good to leave a window for an owner to change them for some flexibility. Leaving them permissionless would introduce security risks. For example, if anybody can set the `min_price` it is possible they will then buy at a price lower then intended by the owner